sop-extras/src/Data/SOP/NonEmpty.hs

```diff
@@ -0,0 +1,175 @@
+
+-- | Type-level non-empty lists
+module Data.SOP.NonEmpty (
+    NonEmpty (..)
+    -- * Proofs
+  , IsNonEmpty (..)
+  , ProofNonEmpty (..)
+  , checkIsNonEmpty
+    -- * Working with 'NonEmpty'
+  , nonEmptyFromList
+  , nonEmptyHead
+  , nonEmptyInit
+  , nonEmptyLast
+  , nonEmptyMapOne
+  , nonEmptyMapTwo
+  , nonEmptyStrictPrefixes
+  , nonEmptyToList
+  ) where
+
+import           Control.Applicative
+import qualified Data.Foldable as Foldable
+import           Data.Kind (Type)
+import           Data.Proxy
+import           Data.SOP.Sing
+
+{-------------------------------------------------------------------------------
+  NonEmpty
+-------------------------------------------------------------------------------}
+
+-- | Non-empty variation on 'AtMost'
+type NonEmpty :: [Type] -> Type -> Type
+data NonEmpty xs a where
+  NonEmptyOne  :: !a -> NonEmpty (x ': xs) a
+  NonEmptyCons :: !a -> !(NonEmpty xs a) -> NonEmpty (x ': xs) a
+
+deriving instance Eq a   => Eq          (NonEmpty xs a)
+deriving instance Show a => Show        (NonEmpty xs a)
+deriving instance           Functor     (NonEmpty xs)
+deriving instance           Foldable    (NonEmpty xs)
+deriving instance           Traversable (NonEmpty xs)
+
+instance (IsNonEmpty xs, SListI xs) => Applicative (NonEmpty xs) where
+  pure :: forall a. a -> NonEmpty xs a
+  pure a = case isNonEmpty (Proxy @xs) of
+             ProofNonEmpty{} -> go shape
+    where
+      go :: forall x xs'. Shape xs' -> NonEmpty (x : xs') a
+      go ShapeNil        = NonEmptyOne a
+      go (ShapeCons xs') = NonEmptyCons a (go xs')
+  (<*>) = go
+    where
+      go :: NonEmpty xs' (a -> b) -> NonEmpty xs' a -> NonEmpty xs' b
+      go (NonEmptyOne  f)    (NonEmptyOne  x)    = NonEmptyOne  (f x)
+      go (NonEmptyCons f _)  (NonEmptyOne  x)    = NonEmptyOne  (f x)
+      go (NonEmptyOne  f)    (NonEmptyCons x _)  = NonEmptyOne  (f x)
+      go (NonEmptyCons f fs) (NonEmptyCons x xs) = NonEmptyCons (f x) (go fs xs)
+
+{-------------------------------------------------------------------------------
+  Working with 'NonEmpty'
+-------------------------------------------------------------------------------}
+
+-- | Analogue of 'head'
+nonEmptyHead :: NonEmpty xs a -> a
+nonEmptyHead (NonEmptyOne  x)   = x
+nonEmptyHead (NonEmptyCons x _) = x
+
+-- | Analogue of 'last'
+nonEmptyLast :: NonEmpty xs a -> a
+nonEmptyLast = snd . nonEmptyInit
+
+-- | Analogue of 'init'
+nonEmptyInit :: NonEmpty xs a -> (Maybe (NonEmpty xs a), a)
+nonEmptyInit (NonEmptyOne  x)    = (Nothing, x)
+nonEmptyInit (NonEmptyCons x xs) =
+    case nonEmptyInit xs of
+      (Nothing  , final) -> (Just (NonEmptyOne  x)     , final)
+      (Just xs' , final) -> (Just (NonEmptyCons x xs') , final)
+
+-- | Build a 'NonEmpty' from a list. Returns 'Nothing' when the list is empty
+-- or when it's longer than @xs@.
+nonEmptyFromList :: forall xs a. SListI xs => [a] -> Maybe (NonEmpty xs a)
+nonEmptyFromList = go (sList @xs)
```

Comment:
A total version of this function, i.e. something returning just `NonEmpty xs a` seems straightforward, since we have the `IsNonEmpty` class. 

Also something for a later PR.